File: tests/wasm/test_volume_zone_oscillator.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadTestData, assertArrayClose, isNaN } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualVzo(close, volume, length, intradaySmoothing, noiseFilter) {
    const alpha = 2 / (length + 1);
    const beta = 1 - alpha;
    const smoothAlpha = 2 / (noiseFilter + 1);
    const smoothBeta = 1 - smoothAlpha;
    const out = new Float64Array(close.length);
    out.fill(NaN);

    let firstValid = -1;
    for (let i = 0; i < volume.length; i++) {
        if (Number.isFinite(volume[i])) {
            firstValid = i;
            break;
        }
    }
    if (firstValid === -1) {
        return out;
    }

    let prevClose = NaN;
    let emaDirection = 0.0;
    let emaTotal = 0.0;
    let smooth = 0.0;
    let smoothValid = false;

    for (let i = firstValid; i < close.length; i++) {
        let raw = NaN;
        if (Number.isFinite(volume[i])) {
            const directed = Number.isFinite(close[i]) && Number.isFinite(prevClose) && close[i] > prevClose
                ? volume[i]
                : -volume[i];
            emaDirection = beta * emaDirection + alpha * directed;
            emaTotal = beta * emaTotal + alpha * volume[i];
            if (emaTotal !== 0) {
                raw = 100 * emaDirection / emaTotal;
            }
        } else if (emaTotal !== 0) {
            raw = 100 * emaDirection / emaTotal;
        }

        if (Number.isFinite(close[i])) {
            prevClose = close[i];
        }

        if (intradaySmoothing) {
            if (Number.isFinite(raw)) {
                smooth = smoothBeta * smooth + smoothAlpha * raw;
                smoothValid = true;
                out[i] = smooth;
            } else if (smoothValid) {
                out[i] = smooth;
            }
        } else {
            out[i] = raw;
        }
    }

    return out;
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('volume_zone_oscillator_js matches manual reference', () => {
    const n = 160;
    const close = new Float64Array(testData.close.slice(0, n));
    const volume = new Float64Array(testData.volume.slice(0, n));
    const result = wasm.volume_zone_oscillator_js(close, volume, 14, true, 4);
    const expected = manualVzo(close, volume, 14, true, 4);
    assertArrayClose(result, expected, 1e-12, 'manual mismatch');
    assert.strictEqual(result.length, n);
    assert.ok(isNaN(result[0]) || Number.isFinite(result[0]));
});

test('volume_zone_oscillator_batch_js first row matches single', () => {
    const n = 128;
    const close = new Float64Array(testData.close.slice(0, n));
    const volume = new Float64Array(testData.volume.slice(0, n));
    const batch = wasm.volume_zone_oscillator_batch_js(close, volume, {
        length_range: [14, 16, 2],
        intraday_smoothing: true,
        noise_filter_range: [4, 4, 0],
    });
    const single = wasm.volume_zone_oscillator_js(close, volume, 14, true, 4);
    assert.strictEqual(batch.rows, 2);
    assert.strictEqual(batch.cols, n);
    assert.strictEqual(batch.combos.length, 2);
    assert.strictEqual(batch.combos[0].length, 14);
    assert.strictEqual(batch.combos[0].intraday_smoothing, true);
    assert.strictEqual(batch.combos[0].noise_filter, 4);
    assertArrayClose(batch.values.slice(0, n), single, 1e-12, 'batch mismatch');
});

test('volume_zone_oscillator_into pointer path matches safe API', () => {
    const n = 128;
    const close = new Float64Array(testData.close.slice(0, n));
    const volume = new Float64Array(testData.volume.slice(0, n));
    const safe = wasm.volume_zone_oscillator_js(close, volume, 14, true, 4);

    const closePtr = wasm.allocate_f64_array(n);
    const volumePtr = wasm.allocate_f64_array(n);
    const outPtr = wasm.volume_zone_oscillator_alloc(n);

    try {
        wasm.write_f64_array(closePtr, close);
        wasm.write_f64_array(volumePtr, volume);
        wasm.volume_zone_oscillator_into(closePtr, volumePtr, outPtr, n, 14, true, 4);
        const flat = wasm.read_f64_array(outPtr, n);
        assertArrayClose(flat, safe, 1e-12, 'pointer mismatch');
    } finally {
        wasm.volume_zone_oscillator_free(outPtr, n);
        wasm.deallocate_f64_array(closePtr);
        wasm.deallocate_f64_array(volumePtr);
    }
});

test('volume_zone_oscillator_batch_into returns row count', () => {
    const n = 96;
    const close = new Float64Array(testData.close.slice(0, n));
    const volume = new Float64Array(testData.volume.slice(0, n));
    const closePtr = wasm.allocate_f64_array(n);
    const volumePtr = wasm.allocate_f64_array(n);
    const rows = 2;
    const outPtr = wasm.volume_zone_oscillator_alloc(rows * n);

    try {
        wasm.write_f64_array(closePtr, close);
        wasm.write_f64_array(volumePtr, volume);
        const returnedRows = wasm.volume_zone_oscillator_batch_into(
            closePtr,
            volumePtr,
            outPtr,
            n,
            14,
            16,
            2,
            true,
            4,
            4,
            0,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.volume_zone_oscillator_free(outPtr, rows * n);
        wasm.deallocate_f64_array(closePtr);
        wasm.deallocate_f64_array(volumePtr);
    }
});

test('volume_zone_oscillator_js rejects invalid length', () => {
    const n = 64;
    const close = new Float64Array(testData.close.slice(0, n));
    const volume = new Float64Array(testData.volume.slice(0, n));
    assert.throws(() => {
        wasm.volume_zone_oscillator_js(close, volume, 1, true, 4);
    }, /Invalid length/);
});

test('volume_zone_oscillator_batch_js rejects invalid config', () => {
    const n = 64;
    const close = new Float64Array(testData.close.slice(0, n));
    const volume = new Float64Array(testData.volume.slice(0, n));
    assert.throws(() => {
        wasm.volume_zone_oscillator_batch_js(close, volume, {
            length_range: [14, 16],
            intraday_smoothing: true,
            noise_filter_range: [4, 4, 0],
        });
    }, /Invalid config/);
});

File: tests/python/test_volume_zone_oscillator.py
import numpy as np
import pytest

from test_utils import load_test_data

try:
    import vector_ta as mp
except ImportError:
    pytest.skip(
        "Python module not built. Run 'maturin develop --features python' first",
        allow_module_level=True,
    )


def manual_vzo(close: np.ndarray, volume: np.ndarray, length: int, intraday_smoothing: bool, noise_filter: int) -> np.ndarray:
    alpha = 2.0 / (length + 1.0)
    beta = 1.0 - alpha
    smooth_alpha = 2.0 / (noise_filter + 1.0)
    smooth_beta = 1.0 - smooth_alpha

    out = np.full(close.shape[0], np.nan, dtype=np.float64)
    valid = np.flatnonzero(np.isfinite(volume))
    if valid.size == 0:
        return out

    first_valid = int(valid[0])
    prev_close = np.nan
    ema_direction = 0.0
    ema_total = 0.0
    smooth = 0.0
    smooth_valid = False

    for i in range(first_valid, close.shape[0]):
        raw = np.nan
        if np.isfinite(volume[i]):
            directed = volume[i] if np.isfinite(close[i]) and np.isfinite(prev_close) and close[i] > prev_close else -volume[i]
            ema_direction = beta * ema_direction + alpha * directed
            ema_total = beta * ema_total + alpha * volume[i]
            if ema_total != 0.0:
                raw = 100.0 * ema_direction / ema_total
        elif ema_total != 0.0:
            raw = 100.0 * ema_direction / ema_total

        if np.isfinite(close[i]):
            prev_close = close[i]

        if intraday_smoothing:
            if np.isfinite(raw):
                smooth = smooth_beta * smooth + smooth_alpha * raw
                smooth_valid = True
                out[i] = smooth
            elif smooth_valid:
                out[i] = smooth
        else:
            out[i] = raw

    return out


class TestVolumeZoneOscillator:
    def test_manual_reference_matches_binding(self):
        data = load_test_data()
        close = np.asarray(data["close"][:160], dtype=np.float64)
        volume = np.asarray(data["volume"][:160], dtype=np.float64)
        expected = manual_vzo(close, volume, 14, True, 4)
        result = mp.volume_zone_oscillator(
            close,
            volume,
            length=14,
            intraday_smoothing=True,
            noise_filter=4,
        )
        np.testing.assert_allclose(result, expected, rtol=0.0, atol=1e-12, equal_nan=True)

    def test_stream_matches_batch(self):
        data = load_test_data()
        close = np.asarray(data["close"][:128], dtype=np.float64)
        volume = np.asarray(data["volume"][:128], dtype=np.float64)
        batch = mp.volume_zone_oscillator(
            close,
            volume,
            length=10,
            intraday_smoothing=False,
            noise_filter=4,
        )
        stream = mp.VolumeZoneOscillatorStream(10, False, 4)
        streamed = np.array([stream.update(float(c), float(v)) for c, v in zip(close, volume)])
        np.testing.assert_allclose(streamed, batch, rtol=0.0, atol=1e-12, equal_nan=True)

    def test_batch_metadata_and_first_row_match_single(self):
        data = load_test_data()
        close = np.asarray(data["close"][:128], dtype=np.float64)
        volume = np.asarray(data["volume"][:128], dtype=np.float64)
        result = mp.volume_zone_oscillator_batch(
            close,
            volume,
            length_range=(14, 16, 2),
            intraday_smoothing=True,
            noise_filter_range=(4, 4, 0),
        )
        assert result["rows"] == 2
        assert result["cols"] == close.shape[0]
        np.testing.assert_array_equal(result["lengths"], np.array([14, 16], dtype=np.uint64))
        np.testing.assert_array_equal(result["noise_filters"], np.array([4, 4], dtype=np.uint64))
        assert result["intraday_smoothing"] == [True, True]

        single = mp.volume_zone_oscillator(
            close,
            volume,
            length=14,
            intraday_smoothing=True,
            noise_filter=4,
        )
        np.testing.assert_allclose(
            result["values"][0],
            single,
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_kernel_parity(self):
        data = load_test_data()
        close = np.asarray(data["close"][:144], dtype=np.float64)
        volume = np.asarray(data["volume"][:144], dtype=np.float64)
        auto = mp.volume_zone_oscillator(close, volume, length=14, intraday_smoothing=True, noise_filter=4)
        scalar = mp.volume_zone_oscillator(
            close,
            volume,
            length=14,
            intraday_smoothing=True,
            noise_filter=4,
            kernel="scalar",
        )
        np.testing.assert_allclose(auto, scalar, rtol=0.0, atol=1e-12, equal_nan=True)

    def test_invalid_length(self):
        data = load_test_data()
        close = np.asarray(data["close"][:32], dtype=np.float64)
        volume = np.asarray(data["volume"][:32], dtype=np.float64)
        with pytest.raises(ValueError, match="Invalid length"):
            mp.volume_zone_oscillator(close, volume, length=1, intraday_smoothing=True, noise_filter=4)

    def test_length_mismatch(self):
        data = load_test_data()
        close = np.asarray(data["close"][:32], dtype=np.float64)
        volume = np.asarray(data["volume"][:31], dtype=np.float64)
        with pytest.raises(ValueError, match="slice length"):
            mp.volume_zone_oscillator(close, volume, length=14, intraday_smoothing=True, noise_filter=4)
